machine/map_state.go

```diff
@@ -0,0 +1,93 @@
+package machine
+
+import (
+	"context"
+	"fmt"
+	"github.com/coinbase/step/jsonpath"
+	"github.com/coinbase/step/utils/to"
+)
+
+type MapState struct {
+	stateStr // Include Defaults
+
+	Type    *string
+	Comment *string `json:",omitempty"`
+
+	Iterator  *StateMachine
+	ItemsPath *jsonpath.Path `json:",omitempty"`
+	Parameters interface{}    `json:",omitempty"`
+
+	MaxConcurrency     *float64       `json:",omitempty"`
+
+	InputPath  *jsonpath.Path `json:",omitempty"`
+	OutputPath *jsonpath.Path `json:",omitempty"`
+	ResultPath *jsonpath.Path `json:",omitempty"`
+
+	Catch []*Catcher `json:",omitempty"`
+	Retry []*Retrier `json:",omitempty"`
+
+	Next *string `json:",omitempty"`
+	End  *bool   `json:",omitempty"`
+}
+
+func (s *MapState) process(ctx context.Context, input interface{}) (interface{}, *string, error) {
+	output, err := s.ItemsPath.GetSlice(input)
+	if err != nil {
+		return input, nextState(s.Next, s.End), err
+	}
+	outputResults := []interface{}{}
```

Comment:
I think we have a type mismatch but I'm not sure how to best adjust `outputResults` here to fix. 
In https://github.com/coinbase/step/blob/master/machine/execution.go#L18 Output is of type `map[string]interface{}{}` but here we made outPutResults `[]interface{}{}`.

If we change `outputResults` to  `map[string]interface{}{}`  I am not sure what the key should be. `s.OutputPath`?

Reply:
Any thoughts on how to best move forward? @grahamjenson

Reply:
> Output is of type `map[string]interface{}{}`

(To be pedantic for a moment: the type is `map[string]interface{}`. Adding another `{}` makes it an empty literal of that type.)

If `execution.Output` is a `map[string]interface{}`, then the thing it's being appended to should be a `[]map[string]interface{}` for maximum typesafety.

Reply:
In other words,

```suggestion
	var outputResults []map[string]interface{}
```

Reply:
Incidentally, `var foo []type` is usually slightly preferable to `foo := []type{}`. The first one sets `foo` to `nil` without allocating any actual slice (which is fine, because `nil` works just the way you expect in `len` and `append` etc). The second one allocates an empty slice, which incurs [a tiny bit of overhead](https://dave.cheney.net/2018/07/12/slices-from-the-ground-up).